File: src/simulation/logistics.js
import { recordEvent } from "./journal.js";
import { resourceTargetStock } from "./demand.js";

const EPSILON = 1e-9;

function quantize(value) {
  return Math.round(value * 1_000_000) / 1_000_000;
}

function incomingAmount(world, cityId, resourceId) {
  return world.shipments
    .filter((shipment) => shipment.to === cityId && shipment.resourceId === resourceId)
    .reduce((total, shipment) => total + shipment.amount, 0);
}

function buildAdjacency(routes, remainingCapacity) {
  const adjacency = new Map();
  for (const route of routes) {
    if ((remainingCapacity.get(route.id) ?? 0) <= EPSILON) continue;
    if (!adjacency.has(route.a)) adjacency.set(route.a, []);
    if (!adjacency.has(route.b)) adjacency.set(route.b, []);
    adjacency.get(route.a).push({ cityId: route.b, route });
    adjacency.get(route.b).push({ cityId: route.a, route });
  }
  for (const neighbors of adjacency.values()) {
    neighbors.sort((left, right) => left.route.id.localeCompare(right.route.id));
  }
  return adjacency;
}

function shortestPath(routes, from, to, remainingCapacity) {
  const adjacency = buildAdjacency(routes, remainingCapacity);
  const best = new Map([[from, { distance: 0, key: "", routeIds: [] }]]);
  const pending = new Set([from]);
  while (pending.size > 0) {
    const current = [...pending].sort((leftId, rightId) => {
      const left = best.get(leftId);
      const right = best.get(rightId);
      return left.distance - right.distance || left.key.localeCompare(right.key) || leftId.localeCompare(rightId);
    })[0];
    pending.delete(current);
    if (current === to) return best.get(current);
    const currentBest = best.get(current);
    for (const edge of adjacency.get(current) ?? []) {
      const candidate = {
        distance: currentBest.distance + edge.route.travelDays,
        key: `${currentBest.key}/${edge.route.id}`,
        routeIds: [...currentBest.routeIds, edge.route.id]
      };
      const known = best.get(edge.cityId);
      if (!known || candidate.distance < known.distance ||
          (candidate.distance === known.distance && candidate.key.localeCompare(known.key) < 0)) {
        best.set(edge.cityId, candidate);
        pending.add(edge.cityId);
      }
    }
  }
  return null;
}

export function deliverShipments(world, telemetry = null) {
  const arrived = world.shipments
    .filter((shipment) => shipment.arrivalDay <= world.day)
    .sort((left, right) => left.id.localeCompare(right.id));
  const arrivedIds = new Set(arrived.map(({ id }) => id));
  world.shipments = world.shipments.filter((shipment) => !arrivedIds.has(shipment.id));
  for (const shipment of arrived) {
    const city = world.cities[shipment.to];
    city.stocks[shipment.resourceId] = quantize(city.stocks[shipment.resourceId] + shipment.amount);
    recordEvent(world, {
      type: "shipment_arrived",
      subjectId: shipment.id,
      causeIds: [shipment.dispatchEventId],
      details: { from: shipment.from, to: shipment.to, resourceId: shipment.resourceId, amount: shipment.amount }
    });
    if (telemetry) telemetry.shipmentsArrived += 1;
  }
}

function outstandingIntent(world, cityId, resourceId, kind) {
  return world.tradeIntents
    .filter((intent) => intent.cityId === cityId && intent.resourceId === resourceId && intent.kind === kind)
    .reduce((sum, intent) => sum + intent.remaining, 0);
}

function createIntent(world, kind, city, resourceId, amount) {
  world.tradeIntents.push({
    id: `intent-${String(world.nextTradeIntentId).padStart(6, "0")}`,
    kind,
    cityId: city.id,
    resourceId,
    amount: quantize(amount),
    remaining: quantize(amount),
    createdDay: world.day,
    availableDay: world.day + 1,
    expiresDay: world.day + 12,
    limitPrice: kind === "demand"
      ? Math.round(city.markets[resourceId].price * 1.18 * 10_000) / 10_000
      : Math.round(city.markets[resourceId].price * 0.9 * 10_000) / 10_000
  });
  world.nextTradeIntentId += 1;
}

function publishLocalIntents(world, content) {
  if (world.day % 3 !== 0) return;
  const resourceIds = content.resources.resources.map((resource) => resource.id).sort();
  for (const cityId of Object.keys(world.cities).sort()) {
    const city = world.cities[cityId];
    for (const resourceId of resourceIds) {
      const target = resourceTargetStock(world, city, resourceId, content);
      const effective = city.stocks[resourceId] + incomingAmount(world, cityId, resourceId);
      const demand = target - effective - outstandingIntent(world, cityId, resourceId, "demand");
      if (target > EPSILON && demand > EPSILON) createIntent(world, "demand", city, resourceId, demand);
      const protectedStock = target * 1.15;
      const offer = city.stocks[resourceId] - protectedStock -
        outstandingIntent(world, cityId, resourceId, "offer");
      if (offer > EPSILON) createIntent(world, "offer", city, resourceId, offer);
    }
  }
}

function matchIntents(world, content, telemetry) {
  const recipeById = new Map(content.recipes.recipes.map((recipe) => [recipe.id, recipe]));
  const remainingCapacity = new Map(world.routes.map((route) => [route.id, route.dailyCapacity]));
  const demands = world.tradeIntents
    .filter((intent) => intent.kind === "demand" && intent.availableDay <= world.day && intent.remaining > EPSILON)
    .sort((left, right) => left.resourceId.localeCompare(right.resourceId) ||
      left.createdDay - right.createdDay || left.id.localeCompare(right.id));

  for (const demand of demands) {
    const offers = world.tradeIntents
      .filter((intent) => intent.kind === "offer" && intent.resourceId === demand.resourceId &&
        intent.cityId !== demand.cityId && intent.availableDay <= world.day && intent.remaining > EPSILON)
      .map((offer) => ({ offer, path: shortestPath(world.routes, offer.cityId, demand.cityId, remainingCapacity) }))
      .filter((candidate) => candidate.path && demand.limitPrice >=
        candidate.offer.limitPrice * (1 + candidate.path.distance * 0.015))
      .sort((left, right) => left.path.distance - right.path.distance ||
        left.path.key.localeCompare(right.path.key) || left.offer.id.localeCompare(right.offer.id));

    for (const { offer, path } of offers) {
      if (demand.remaining <= EPSILON) break;
      const source = world.cities[offer.cityId];
      const target = resourceTargetStock(world, source, demand.resourceId, content);
      const actualSurplus = Math.max(0, source.stocks[demand.resourceId] - target * 1.05);
      const pathCapacity = Math.min(...path.routeIds.map((routeId) => remainingCapacity.get(routeId)));
      const amount = quantize(Math.min(demand.remaining, offer.remaining, actualSurplus, pathCapacity));
      if (amount <= EPSILON) continue;
      source.stocks[demand.resourceId] = quantize(source.stocks[demand.resourceId] - amount);
      demand.remaining = quantize(demand.remaining - amount);
      offer.remaining = quantize(offer.remaining - amount);
      for (const routeId of path.routeIds) {
        remainingCapacity.set(routeId, quantize(remainingCapacity.get(routeId) - amount));
      }

      const shipmentId = `shipment-${String(world.nextShipmentId).padStart(6, "0")}`;
      world.nextShipmentId += 1;
      const dispatchEvent = recordEvent(world, {
        type: "shipment_dispatched",
        subjectId: shipmentId,
        causeIds: source.resourceSignals[demand.resourceId] ? [source.resourceSignals[demand.resourceId]] : [],
        details: {
          from: offer.cityId,
          to: demand.cityId,
          resourceId: demand.resourceId,
          amount,
          routeIds: path.routeIds,
          offerIntentId: offer.id,
          demandIntentId: demand.id,
          unitPrice: Math.round(((offer.limitPrice + demand.limitPrice) / 2) * 10_000) / 10_000,
          arrivalDay: world.day + path.distance
        }
      });
      world.shipments.push({
        id: shipmentId,
        from: offer.cityId,
        to: demand.cityId,
        resourceId: demand.resourceId,
        amount,
        routeIds: path.routeIds,
        departureDay: world.day,
        arrivalDay: world.day + path.distance,
        dispatchEventId: dispatchEvent.id
      });
      telemetry.shipmentsDispatched += 1;
    }
    if (demand.remaining > EPSILON && !demand.shortfallEventId) {
      const causeIds = Object.values(world.cities).flatMap((city) => city.industries
        .filter((industry) => Object.hasOwn(recipeById.get(industry.recipeId).outputs, demand.resourceId))
        .map((industry) => industry.constraintEventId)
        .filter(Boolean));
      const shortfall = recordEvent(world, {
        type: "trade_shortfall",
        subjectId: demand.id,
        causeIds,
        details: {
          cityId: demand.cityId,
          resourceId: demand.resourceId,
          missingAmount: demand.remaining
        }
      });
      demand.shortfallEventId = shortfall.id;
      world.cities[demand.cityId].resourceSignals[demand.resourceId] = shortfall.id;
    }
  }
  world.shipments.sort((left, right) => left.id.localeCompare(right.id));
}

export function planShipments(world, content, telemetry = { shipmentsDispatched: 0 }) {
  world.tradeIntents = world.tradeIntents.filter((intent) =>
    intent.remaining > EPSILON && intent.expiresDay > world.day
  );
  publishLocalIntents(world, content);
  matchIntents(world, content, telemetry);
  world.tradeIntents = world.tradeIntents.filter((intent) =>
    intent.remaining > EPSILON && intent.expiresDay > world.day
  );
  world.tradeIntents.sort((left, right) => left.id.localeCompare(right.id));
}

File: src/WorldGen.Core/Simulation/Logistics.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using WorldGen.Core.Content;

namespace WorldGen.Core.Simulation;

public static class Logistics
{
    private const double Epsilon = 1e-9;

    public static void Deliver(WorldState world, DailyTelemetry? telemetry = null)
    {
        var arrived = world.Shipments.Where(shipment => shipment.ArrivalDay <= world.Day).OrderBy(shipment => shipment.Id, StringComparer.Ordinal).ToArray();
        var ids = arrived.Select(item => item.Id).ToHashSet(StringComparer.Ordinal);
        world.Shipments = world.Shipments.Where(shipment => !ids.Contains(shipment.Id)).ToList();
        foreach (var shipment in arrived)
        {
            var city = world.Cities[shipment.To];
            city.Stocks[shipment.ResourceId] = SimulationMath.Quantize(city.Stocks[shipment.ResourceId] + shipment.Amount);
            Journal.Record(world, "shipment_arrived", shipment.Id, [shipment.DispatchEventId],
                new JsonObject { ["from"] = shipment.From, ["to"] = shipment.To, ["resourceId"] = shipment.ResourceId, ["amount"] = shipment.Amount });
            if (telemetry is not null) telemetry.ShipmentsArrived++;
        }
    }

    public static void Plan(WorldState world, ContentCatalog content, DailyTelemetry telemetry)
    {
        world.TradeIntents = world.TradeIntents.Where(intent => intent.Remaining > Epsilon && intent.ExpiresDay > world.Day).ToList();
        Publish(world, content);
        Match(world, content, telemetry);
        world.TradeIntents = world.TradeIntents.Where(intent => intent.Remaining > Epsilon && intent.ExpiresDay > world.Day)
            .OrderBy(intent => intent.Id, StringComparer.Ordinal).ToList();
    }

    private static void Publish(WorldState world, ContentCatalog content)
    {
        if (world.Day % 3 != 0) return;
        foreach (var cityId in world.Cities.Keys.Order(StringComparer.Ordinal))
        foreach (var resourceId in content.Resources.Resources.Select(resource => resource.Id).Order(StringComparer.Ordinal))
        {
            var city = world.Cities[cityId];
            var target = NeedsAndDemand.ResourceTargetStock(world, city, resourceId, content);
            var incoming = world.Shipments.Where(item => item.To == cityId && item.ResourceId == resourceId).Sum(item => item.Amount);
            var effective = city.Stocks[resourceId] + incoming;
            var demand = target - effective - Outstanding(world, cityId, resourceId, "demand");
            if (target > Epsilon && demand > Epsilon) CreateIntent(world, "demand", city, resourceId, demand);
            var offer = city.Stocks[resourceId] - target * 1.15 - Outstanding(world, cityId, resourceId, "offer");
            if (offer > Epsilon) CreateIntent(world, "offer", city, resourceId, offer);
        }
    }

    private static double Outstanding(WorldState world, string cityId, string resourceId, string kind) => world.TradeIntents
        .Where(intent => intent.CityId == cityId && intent.ResourceId == resourceId && intent.Kind == kind).Sum(intent => intent.Remaining);

    private static void CreateIntent(WorldState world, string kind, CityState city, string resourceId, double amount)
    {
        world.TradeIntents.Add(new TradeIntentState
        {
            Id = $"intent-{world.NextTradeIntentId:000000}", Kind = kind, CityId = city.Id, ResourceId = resourceId,
            Amount = SimulationMath.Quantize(amount), Remaining = SimulationMath.Quantize(amount), CreatedDay = world.Day,
            AvailableDay = world.Day + 1, ExpiresDay = world.Day + 12,
            LimitPrice = SimulationMath.Round(city.Markets[resourceId].Price * (kind == "demand" ? 1.18 : 0.9), 10_000)
        });
        world.NextTradeIntentId++;
    }

    private static void Match(WorldState world, ContentCatalog content, DailyTelemetry telemetry)
    {
        var recipes = content.Recipes.Recipes.ToDictionary(recipe => recipe.Id, StringComparer.Ordinal);
        var capacity = world.Routes.ToDictionary(route => route.Id, route => route.DailyCapacity, StringComparer.Ordinal);
        var demands = world.TradeIntents.Where(intent => intent.Kind == "demand" && intent.AvailableDay <= world.Day && intent.Remaining > Epsilon)
            .OrderBy(intent => intent.ResourceId, StringComparer.Ordinal).ThenBy(intent => intent.CreatedDay).ThenBy(intent => intent.Id, StringComparer.Ordinal).ToArray();
        foreach (var demand in demands)
        {
            var offers = world.TradeIntents.Where(intent => intent.Kind == "offer" && intent.ResourceId == demand.ResourceId &&
                    intent.CityId != demand.CityId && intent.AvailableDay <= world.Day && intent.Remaining > Epsilon)
                .Select(offer => (Offer: offer, Path: ShortestPath(world.Routes, offer.CityId, demand.CityId, capacity)))
                .Where(candidate => candidate.Path is not null && demand.LimitPrice >= candidate.Offer.LimitPrice * (1 + candidate.Path.Distance * 0.015))
                .OrderBy(candidate => candidate.Path!.Distance).ThenBy(candidate => candidate.Path!.Key, StringComparer.Ordinal)
                .ThenBy(candidate => candidate.Offer.Id, StringComparer.Ordinal).ToArray();
            foreach (var candidate in offers)
            {
                if (demand.Remaining <= Epsilon) break;
                var offer = candidate.Offer; var path = candidate.Path!; var source = world.Cities[offer.CityId];
                var target = NeedsAndDemand.ResourceTargetStock(world, source, demand.ResourceId, content);
                var surplus = Math.Max(0, source.Stocks[demand.ResourceId] - target * 1.05);
                var pathCapacity = path.RouteIds.Min(id => capacity[id]);
                var amount = SimulationMath.Quantize(Math.Min(Math.Min(demand.Remaining, offer.Remaining), Math.Min(surplus, pathCapacity)));
                if (amount <= Epsilon) continue;
                source.Stocks[demand.ResourceId] = SimulationMath.Quantize(source.Stocks[demand.ResourceId] - amount);
                demand.Remaining = SimulationMath.Quantize(demand.Remaining - amount); offer.Remaining = SimulationMath.Quantize(offer.Remaining - amount);
                foreach (var routeId in path.RouteIds) capacity[routeId] = SimulationMath.Quantize(capacity[routeId] - amount);
                var shipmentId = $"shipment-{world.NextShipmentId:000000}"; world.NextShipmentId++;
                var evt = Journal.Record(world, "shipment_dispatched", shipmentId,
                    source.ResourceSignals.TryGetValue(demand.ResourceId, out var cause) ? [cause] : [],
                    new JsonObject { ["from"] = offer.CityId, ["to"] = demand.CityId, ["resourceId"] = demand.ResourceId,
                        ["amount"] = amount, ["routeIds"] = JsonSerializer.SerializeToNode(path.RouteIds), ["offerIntentId"] = offer.Id,
                        ["demandIntentId"] = demand.Id, ["unitPrice"] = SimulationMath.Round((offer.LimitPrice + demand.LimitPrice) / 2, 10_000),
                        ["arrivalDay"] = world.Day + path.Distance });
                world.Shipments.Add(new ShipmentState(shipmentId, offer.CityId, demand.CityId, demand.ResourceId, amount,
                    path.RouteIds, world.Day, world.Day + path.Distance, evt.Id));
                telemetry.ShipmentsDispatched++;
            }
            if (demand.Remaining > Epsilon && demand.ShortfallEventId is null)
            {
                var causes = world.Cities.Values.SelectMany(city => city.Industries
                    .Where(industry => recipes[industry.RecipeId].Outputs.ContainsKey(demand.ResourceId))
                    .Select(industry => industry.ConstraintEventId)).Where(id => id is not null);
                var evt = Journal.Record(world, "trade_shortfall", demand.Id, causes,
                    new JsonObject { ["cityId"] = demand.CityId, ["resourceId"] = demand.ResourceId, ["missingAmount"] = demand.Remaining });
                demand.ShortfallEventId = evt.Id; world.Cities[demand.CityId].ResourceSignals[demand.ResourceId] = evt.Id;
            }
        }
        world.Shipments = world.Shipments.OrderBy(shipment => shipment.Id, StringComparer.Ordinal).ToList();
    }

    private static PathState? ShortestPath(List<RouteState> routes, string from, string to, Dictionary<string, double> capacity)
    {
        var adjacency = new Dictionary<string, List<(string CityId, RouteState Route)>>(StringComparer.Ordinal);
        foreach (var route in routes.Where(route => capacity.GetValueOrDefault(route.Id) > Epsilon))
        {
            if (!adjacency.TryGetValue(route.A, out var a)) adjacency[route.A] = a = [];
            if (!adjacency.TryGetValue(route.B, out var b)) adjacency[route.B] = b = [];
            a.Add((route.B, route)); b.Add((route.A, route));
        }
        foreach (var list in adjacency.Values) list.Sort((a, b) => StringComparer.Ordinal.Compare(a.Route.Id, b.Route.Id));
        var best = new Dictionary<string, PathState>(StringComparer.Ordinal) { [from] = new(0, "", []) };
        var pending = new HashSet<string>(StringComparer.Ordinal) { from };
        while (pending.Count > 0)
        {
            var current = pending.OrderBy(id => best[id].Distance).ThenBy(id => best[id].Key, StringComparer.Ordinal).ThenBy(id => id, StringComparer.Ordinal).First();
            pending.Remove(current); if (current == to) return best[current];
            if (!adjacency.TryGetValue(current, out var edges)) continue;
            foreach (var edge in edges)
            {
                var known = best.GetValueOrDefault(edge.CityId); var currentBest = best[current];
                var candidate = new PathState(currentBest.Distance + edge.Route.TravelDays, $"{currentBest.Key}/{edge.Route.Id}", [.. currentBest.RouteIds, edge.Route.Id]);
                if (known is null || candidate.Distance < known.Distance || candidate.Distance == known.Distance && StringComparer.Ordinal.Compare(candidate.Key, known.Key) < 0)
                { best[edge.CityId] = candidate; pending.Add(edge.CityId); }
            }
        }
        return null;
    }

    private sealed record PathState(int Distance, string Key, List<string> RouteIds);
}
